signer/tests/regtest/mod.rs

```diff
@@ -0,0 +1,363 @@
+            input: vec![TxIn {
+                previous_output: OutPoint::new(utxo.txid, utxo.vout),
+                sequence: Sequence::ZERO,
+                script_sig: ScriptBuf::new(),
+                witness: Witness::new(),
+            }],
+            output: vec![
+                TxOut {
+                    value: Amount::from_sat(amount),
+                    script_pubkey: address.script_pubkey(),
+                },
+                TxOut {
+                    value: utxo.amount.unchecked_sub(Amount::from_sat(amount + fee)),
+                    script_pubkey: self.address.script_pubkey(),
+                },
+            ],
+        };
+
+        let input_index = 0;
+        let tx = match self.address.address_type().unwrap() {
+            AddressType::P2wpkh => p2wpkh_sign_transaction(tx, input_index, &utxo, &self.keypair),
+            AddressType::P2tr => {
+                let leaf = Either::Left(None);
+                let (mut tx, signature) =
+                    p2tr_signature(tx, input_index, &[utxo], self.keypair, leaf);
+                tx.input[input_index].witness = Witness::p2tr_key_spend(&signature);
+                tx
+            }
+            _ => unimplemented!(),
+        };
+        rpc.send_raw_transaction(&tx).unwrap();
+    }
+}
+
+pub trait Utxo {
+    fn amount(&self) -> Amount;
+    fn script_pubkey(&self) -> &ScriptBuf;
+    fn to_tx_out(&self) -> TxOut {
+        TxOut {
+            value: self.amount(),
+            script_pubkey: self.script_pubkey().clone(),
+        }
+    }
+}
+
+impl Utxo for ListUnspentResultEntry {
+    fn amount(&self) -> Amount {
+        self.amount
+    }
+
+    fn script_pubkey(&self) -> &ScriptBuf {
+        &self.script_pub_key
+    }
+}
+
+impl Utxo for TxOut {
+    fn amount(&self) -> Amount {
+        self.value
+    }
+
+    fn script_pubkey(&self) -> &ScriptBuf {
+        &self.script_pubkey
+    }
+}
+
+pub fn p2wpkh_sign_transaction<U>(
+    tx: Transaction,
+    input_index: usize,
+    utxo: &U,
+    keys: &Keypair,
+) -> Transaction
+where
+    U: Utxo,
+{
+    let sighash_type = EcdsaSighashType::All;
+    let mut sighasher = SighashCache::new(tx);
+    let sighash = sighasher
+        .p2wpkh_signature_hash(
+            input_index,
+            utxo.script_pubkey().as_script(),
+            utxo.amount(),
+            sighash_type,
+        )
+        .expect("failed to create sighash");
+
+    let msg = Message::from(sighash);
+    let signature = SECP256K1.sign_ecdsa(&msg, &keys.secret_key());
+
+    let signature = bitcoin::ecdsa::Signature { signature, sighash_type };
+    *sighasher.witness_mut(input_index).unwrap() = Witness::p2wpkh(&signature, &keys.public_key());
+
+    sighasher.into_transaction()
+}
+
+/// The enum `Either` with variants `Left` and `Right` is a general purpose
+/// sum type with two cases.
+pub enum Either<L, R> {
+    /// A value of type `L`.
+    Left(L),
+    /// A value of type `R`.
+    Right(R),
+}
```

Comment:
Might be overkill here, but have you considered using the [Either](https://docs.rs/either/latest/either/) crate for this?

Reply:
This is actually copied from that crate. I did consider importing it and thought it would be overkill. If we need it for something else I'm okay using the crate.

Reply:
Hahah I thought it looked familiar :)